### ISSUE TITLE
Refactoring Of Abbreviation Feature

### ISSUE THREAD

**Issue by hediet** [leanprover/vscode-lean#239]
Currently, the abbreviation feature (hover + completion) is all implemented in `input.ts`.

I think it might make sense to clean up that file a little bit and to split it up.
I'm working on a PR right now that presents some ideas. This issue can be used to discuss stuff before I submit the PR in the next few days.

What do you think?

After that restructuring, I want to improve the abbreviation feature to give something back for all the lean-help I got on zulip.
Ideas include:
* Autocompletion for abbreviations
* Live preview of the abbreviated symbol

**Comment by hediet** [comment]
Also, the current range tracking algorithm has some issues with multi cursor support:

![ee62cf9a-b7c4-4624-a034-d84c8ca4685d](https://user-images.githubusercontent.com/2931520/103302694-806ec600-4a04-11eb-8649-5f421fc7b601.gif)

**Comment by bryangingechen** [comment]
Multi-cursor support for unicode translations was added somewhat recently in #219 by @alexpeattie.

**Comment by hediet** [comment]
I guess the current implementation assumes that tracked ranges don't get shifted.
However, this can actually happen when multiple ranges are in a single line.
Thanks for linking the PR!

**Comment by alexpeattie** [comment]
Yes I think a clean refactor of this would be great @hediet 🙂! I think it would be great to do away with all of the "hacky replacement" stuff if possible, and there are other suboptimal things in there, like needing separate handlers for `convert` and `convertWithNewline`

**Comment by hediet** [comment]
I think it would be nice if
1) abbreviations could be replaced as soon as they are not the prefix of any longer abbreviation anymore (this would be a generalization of extending a range with a whitespace character). This could also solve the `\\` case with `\` being an abbreviation for `\`.
2) abbreviations can have cursor placeholders where the cursor is placed after the abbreviation is unfolded. This would fix the hacky stuff.

**Comment by hediet** [comment]
What are the usecases for `convert` and `convertWithNewline`? Why wouldn't you just press enter or the arrow key?

**Comment by alexpeattie** [comment]
Tab and Enter are bound to `convert` and `convertWithNewline` respectively, where the latter differs in that it inserts a `\n`. The arrow keys don't use either, but rather the `onSelectionChanged` event ☺️

**Comment by hediet** [comment]
But why isn't convert and convertWithNewline handled by a text change event?

**Comment by alexpeattie** [comment]
Yeah it was easiest to set it up that way because of the range tracking problem (which is obviously not 100% fixed anyway as you rightly point out) - e.g. the Enter key would be pressed, which then inserts a newline which would mess up the tracked range positions. So using the custom command just delays inserting the newline until after the range conversion is done

**Comment by hediet** [comment]
What do you think of renaming `leader` to `abbreviation character`? `\` in `Hello\nWorld` is called escape character and it starts an escape sequence. `\` in `\alpha` starts an abbreviation sequence, so I think it makes sense to call it `abbreviation character`.
